File: ex/clean/src/main/kotlin/com/ktb/shop/usecases/port/CustomerGateway.kt
package com.ktb.shop.usecases.port

import com.ktb.shop.entities.Customer
import com.ktb.shop.entities.Email

// OUTPUT port (a "gateway" in Clean terms): the use-case layer owns this contract and
// calls outward through it. The implementation lives in the adapters layer.
interface CustomerGateway {
    fun findByEmail(email: Email): Customer?
    fun save(customer: Customer)
}

File: ex/clean-go/internal/usecases/port/customer_gateway.go
package port

import (
	"context"

	"github.com/ktb/clean-shop/internal/entities"
)

// CustomerGateway is an OUTPUT port (a "gateway" in Clean terms): the use-case
// layer owns this contract and calls outward through it. The implementation
// lives in the adapters layer. A nil *Customer with a nil error means "not found".
type CustomerGateway interface {
	FindByEmail(ctx context.Context, email entities.Email) (*entities.Customer, error)
	Save(ctx context.Context, customer entities.Customer) error
}
